crates/red_knot_python_semantic/src/types.rs

```diff
@@ -600,11 +599,16 @@ impl<'db> Type<'db> {
 
     /// Return true if this type is a [subtype of] type `target`.
     ///
+    /// This method returns `false` if either `self` or `other` is not fully static.
+    ///
     /// [subtype of]: https://typing.readthedocs.io/en/latest/spec/concepts.html#subtype-supertype-and-type-equivalence
     pub(crate) fn is_subtype_of(self, db: &'db dyn Db, target: Type<'db>) -> bool {
         if self.is_equivalent_to(db, target) {
             return true;
         }
+        if !self.is_fully_static(db) || !target.is_fully_static(db) {
+            return false;
+        }
```

Comment:
Benchmarks CI job passed with a green check, but there is no comment from codspeed? Is that normal?

If benchmarks are OK I am fine with doing this for now and considering optimizations if it shows up as a performance issue.

Another thought I have is that if we defined a cheaper `is_fully_static_shallow` which didn't descend into unions, intersections, tuples, etc, and only checked that here, would that still give the right behavior? Because if a complex type has a non-fully-static member, wouldn't we eventually hit that member in the actual subtype check and still return `false`? It may be more complex than this for intersections... also it's not entirely self-evident that this will be faster; if a high proportion of types are not fully static, it may be cheaper to rule them out quickly than to go into the actual subtype check.

Reply:
> Benchmarks CI job passed with a green check, but there is no comment from codspeed? Is that normal?

Yes, this is normal. Codspeed only comments if there's at least a 4% regression (or if there was one in a previous push)

Reply:
> Another thought I have is that if we defined a cheaper `is_fully_static_shallow` which didn't descend into unions, intersections, tuples, etc, and only checked that here, would that still give the right behavior?

Hm. It's late here and I should probably sleep first, but I think that won't work (without additional checks below). In the example above, the `int | Any` union would pass the shallow check (there is no gradual form at the top level). And then we would check if `int` is a subtype of `int | Any` using
```rs
            (ty, Type::Union(union)) => union
                .elements(db)
                .iter()
                .any(|&elem_ty| ty.is_subtype_of(db, elem_ty)),
```
which would return true because `int <: int`. We wouldn't even look at `Any` in the traversal due to short-circuiting. We *could* add another `.iter().all(|elem_ty| elem_ty.is_fully_static_shallow())` test here and that would probably be sound, but that would sort of defeat the purpose? There would be a benefit for very deep nested types, but I'm not sure it's worth the additional complexity?

Reply:
> Do you just go to https:://codspeed.io/astral-sh/ruff and find the PR there?

This is what I do. Go to [branches](https://codspeed.io/astral-sh/ruff/branches) and then select it there.

Reply:
Yes, you're right. I think if we find that we need to improve this, the thing to try might be to see if we can steal a bit from `UnionType`, `IntersectionType` etc, and track fully-static-ness eagerly as we construct the type, so that the read check can be shallow and cheap.

Reply:
> Thank you! How did you get that link to the actual results, without the comment?


You can click on details next to the codspeed check. Then there's another link to the benchmark results. 

![image](https://github.com/user-attachments/assets/6929631e-48c2-4e94-ab5d-e2c71f4dbfc1)